File: C#/TodoAppAPI/DTOs/BoardCardFilterRequest.cs
using TodoAppAPI.Models;

namespace TodoAppAPI.DTOs
{
    public class BoardCardFilterRequest
    {
        public string? Keyword { get; set; }
        public bool NoMembers { get; set; }
        public bool AssignedToMe { get; set; }
        public List<string> MemberUIds { get; set; } = new();
        public string? CompletionStatus { get; set; }
        public List<string> DueDateFilters { get; set; } = new();
        public bool NoLabels { get; set; }
        public List<string> LabelUIds { get; set; } = new();
        public List<BoardCardLabelFilterGroupRequest> SelectedLabelGroups { get; set; } = new();
        public string MatchMode { get; set; } = BoardCardFilterValues.MatchExact;
    }

    public class BoardCardLabelFilterGroupRequest
    {
        public List<string> CardLabelUIds { get; set; } = new();
    }

    public static class BoardCardFilterValues
    {
        public const string MatchExact = "exact";
        public const string MatchAny = "any";
        public const string CompletionCompleted = "completed";
        public const string CompletionIncomplete = "incomplete";
        public const string DueOverdue = "overdue";
        public const string DueNoDate = "no_due_date";
        public const string DueNextWeek = "next_week";
        public const string DueNextMonth = "next_month";
    }

    public enum BoardCardFilterResultStatus
    {
        Success,
        BadRequest,
        Forbidden,
        NotFound
    }

    public class BoardCardFilterResult
    {
        public BoardCardFilterResultStatus Status { get; init; }
        public string? Message { get; init; }
        public List<Card> Cards { get; init; } = new();

        public static BoardCardFilterResult Success(List<Card> cards) =>
            new() { Status = BoardCardFilterResultStatus.Success, Cards = cards };

        public static BoardCardFilterResult BadRequest(string message) =>
            new() { Status = BoardCardFilterResultStatus.BadRequest, Message = message };

        public static BoardCardFilterResult Forbidden(string message) =>
            new() { Status = BoardCardFilterResultStatus.Forbidden, Message = message };

        public static BoardCardFilterResult NotFound(string message) =>
            new() { Status = BoardCardFilterResultStatus.NotFound, Message = message };
    }
}

File: Flutter/trellon_flutter/lib/features/board/data/models/board_card_filter_request.dart
class BoardCardFilterRequest {
  final String keyword;
  final bool noMembers;
  final bool assignedToMe;
  final Set<String> memberUIds;
  final String? completionStatus;
  final Set<String> dueDateFilters;
  final bool noLabels;
  final Set<String> labelUIds;
  final List<BoardCardLabelFilterGroupRequest> selectedLabelGroups;
  final String matchMode;

  const BoardCardFilterRequest({
    this.keyword = '',
    this.noMembers = false,
    this.assignedToMe = false,
    this.memberUIds = const {},
    this.completionStatus,
    this.dueDateFilters = const {},
    this.noLabels = false,
    this.labelUIds = const {},
    this.selectedLabelGroups = const [],
    this.matchMode = 'exact',
  });

  bool get isActive =>
      keyword.trim().isNotEmpty ||
      noMembers ||
      assignedToMe ||
      memberUIds.isNotEmpty ||
      completionStatus != null ||
      dueDateFilters.isNotEmpty ||
      noLabels ||
      labelUIds.isNotEmpty ||
      selectedLabelGroups.isNotEmpty;

  Map<String, dynamic> toJson() {
    return {
      'keyword': keyword.trim(),
      'noMembers': noMembers,
      'assignedToMe': assignedToMe,
      'memberUIds': memberUIds.toList(),
      'completionStatus': completionStatus,
      'dueDateFilters': dueDateFilters.toList(),
      'noLabels': noLabels,
      'labelUIds': labelUIds.toList(),
      'selectedLabelGroups': selectedLabelGroups
          .map((group) => group.toJson())
          .toList(growable: false),
      'matchMode': matchMode,
    };
  }
}

class BoardCardLabelFilterGroupRequest {
  final List<String> cardLabelUIds;

  const BoardCardLabelFilterGroupRequest({required this.cardLabelUIds});

  Map<String, dynamic> toJson() {
    return {'cardLabelUIds': cardLabelUIds};
  }

  @override
  bool operator ==(Object other) {
    return other is BoardCardLabelFilterGroupRequest &&
        _listEquals(cardLabelUIds, other.cardLabelUIds);
  }

  @override
  int get hashCode => Object.hashAll(cardLabelUIds);

  static bool _listEquals(List<String> a, List<String> b) {
    if (a.length != b.length) return false;
    for (var i = 0; i < a.length; i++) {
      if (a[i] != b[i]) return false;
    }
    return true;
  }
}
